core/lib/zksync_core/src/state_keeper/io/event_indexes_migration.rs

```diff
@@ -0,0 +1,395 @@
+        let chunk_end = last_miniblock.min(chunk_start + chunk_size - 1);
+        let chunk = chunk_start..=chunk_end;
+
+        let mut storage = pool.access_storage().await?;
+        let is_chunk_migrated = are_event_indexes_migrated(&mut storage, chunk.clone()).await?;
+
+        if is_chunk_migrated {
+            tracing::debug!("Event indexes are migrated for chunk {chunk:?}");
+        } else {
+            tracing::debug!("Migrating event indexes for miniblocks chunk {chunk:?}");
+
+            let rows_affected = storage
+                .events_dal()
+                .assign_indexes_without_eth_transfer(chunk.clone())
+                .await
+                .with_context(|| {
+                    format!("Failed migrating events in miniblocks, chunk {chunk:?}")
+                })?;
+            tracing::debug!("Migrated {rows_affected} events in chunk {chunk:?}");
+            events_affected += rows_affected;
+        }
+        drop(storage);
+
+        if *stop_receiver.borrow() {
+            tracing::info!("Stop signal received; event index migration shutting down");
+            return Ok(MigrationOutput { events_affected });
+        }
+        chunk_start = chunk_end + 1;
+
+        if !is_chunk_migrated {
+            tokio::time::sleep(sleep_interval).await;
+        }
+    }
+
+    Ok(MigrationOutput { events_affected })
+}
+
+#[allow(deprecated)]
+async fn are_event_indexes_migrated(
+    storage: &mut StorageProcessor<'_>,
+    range: ops::RangeInclusive<MiniblockNumber>,
+) -> anyhow::Result<bool> {
+    storage
+        .events_dal()
+        .are_event_indexes_migrated(range.clone())
+        .await
+        .with_context(|| {
+            format!(
+                "Failed getting event indexes for miniblocks in range #{}..=#{}",
+                range.start().0,
+                range.end().0
+            )
+        })
+}
+
+#[cfg(test)]
+mod tests {
+    use multivm::zk_evm_1_3_1::ethereum_types::H256;
+    use test_casing::test_casing;
+    use zksync_system_constants::{L2_ETH_TOKEN_ADDRESS, TRANSFER_EVENT_TOPIC};
+    use zksync_types::{
+        api::GetLogsFilter, tx::IncludedTxLocation, Address, L1BatchNumber, ProtocolVersion,
+        VmEvent,
+    };
+
+    use super::*;
+    use crate::utils::testonly::create_miniblock;
+
+    async fn store_events(
+        storage: &mut StorageProcessor<'_>,
+        miniblock_number: u32,
+        start_idx: u32,
+    ) -> anyhow::Result<(IncludedTxLocation, Vec<VmEvent>)> {
+        let new_miniblock = create_miniblock(miniblock_number);
+        storage
+            .blocks_dal()
+            .insert_miniblock(&new_miniblock)
+            .await?;
+        let tx_location = IncludedTxLocation {
+            tx_hash: H256::repeat_byte(1),
+            tx_index_in_miniblock: 0,
+            tx_initiator_address: Address::repeat_byte(2),
+        };
+        let events = vec![
+            // Matches address, doesn't match topics
```

Comment:
Nit: These comments don't make much sense here, and I'd argue that it would be sufficient to retain only the last 3 events (i.e., ones related to filtering logic).